crates/cairo-profiler/src/profile_viewer.rs

```diff
@@ -0,0 +1,167 @@
+
+    let mut profile_map = HashMap::<String, FunctionProfile>::new();
+
+    let location_map: HashMap<u64, &Location> = profile
+        .location
+        .iter()
+        .map(|location| (location.id, location))
+        .collect();
+
+    let function_map: HashMap<u64, &Function> = profile
+        .function
+        .iter()
+        .map(|function| (function.id, function))
+        .collect();
+
+    for sample in &profile.sample {
+        let sample_value = sample.value[sample_type_idx];
+
+        sample
+            .location_id
+            .iter()
+            .filter_map(|&loc_id| {
+                let line = location_map.get(&loc_id)?.line.first()?;
+                let function = function_map.get(&line.function_id)?;
+                Some(
+                    &profile.string_table[usize::try_from(function.name)
+                        .expect("Overflow while converting function id to usize")],
+                )
+            })
+            .enumerate()
+            .for_each(|(idx, function_name)| {
+                let entry = profile_map.entry(function_name.clone()).or_default();
+                entry.cumulative += sample_value;
+                if idx == 0 {
+                    entry.flat += sample_value;
+                }
+            });
+    }
+
+    let total_resource_count = profile_map
+        .values()
+        .max_by_key(|function| function.cumulative)
+        .map(|function_profile| function_profile.cumulative)
+        .context("Failed to obtain total resource count from cumulative stats")?;
+
+    // sum_p depends on the correct order of data
+    let mut sorted_profile_map: Vec<(String, FunctionProfile)> = profile_map.into_iter().collect();
+    sorted_profile_map.sort_by(|a, b| b.1.flat.cmp(&a.1.flat));
+
+    let mut sum_p: f64 = 0.0;
+
+    for (_, profile) in &mut sorted_profile_map {
+        sum_p += profile.flat as f64;
+
+        profile.flat_p = (profile.flat as f64 / total_resource_count as f64) * 100.0;
+        profile.cumulative_p = (profile.cumulative as f64 / total_resource_count as f64) * 100.0;
+        profile.sum_p = (sum_p / total_resource_count as f64) * 100.0;
+    }
+
+    Ok(sorted_profile_map)
+}
+
+pub fn get_samples(profile: &Profile) -> Vec<&str> {
+    profile
+        .sample_type
+        .iter()
+        .map(|sample| {
+            profile.string_table[usize::try_from(sample.unit)
+                .expect("Overflow while converting samples id to usize")]
+            .as_str()
+        })
+        .collect()
+}
+
+pub fn print_profile(profile: &Profile, sample: &str, limit: NonZeroUsize) -> Result<()> {
+    let data = get_profile_data(profile, sample).context("Failed to get data from profile")?;
+
+    let total_resource_count = data
+        .iter()
+        .max_by_key(|(_, profile)| profile.cumulative)
+        .map(|(_, profile)| profile.cumulative)
+        .context("Failed to obtain total resource count from profile data")?;
+
+    let profile_length = &data.len();
```

Comment:
[nit]

```rs
    let profile_length = data.len();
    let effective_limit = std::cmp::min(limit.get(), profile_length);
    let sliced = data.iter().take(effective_limit).collect::<Vec<_>>();
```

we can compare values not references. This way we don't need binding